### PR TITLE
Grpc

### PR DESCRIPTION
**Pull request by iliat**
I have refactored the code a lot to support both Genomics API and GRPC objects.
Since the objects are similar but have no common ancestor and because I did not want to compromise performance with wrapper classes, I opted for generics.
I split most of the existing classes into interface and base implementation and kept as much functionality as possible in base classes and pushes API specific and GRPC specific code into classes in /api and /grpc folders. Because there is no duck typing in Java there is still some duplication between these two but I think its ok since I expect to eventually only leave grpc version around.

There are subtle differences in the conversion of Read -> SAMRecord/Header since GRPC versions tend not to return nulls but rather default values when string fields are not set and CigarUnit is an enum as opposed to String.

I know we moved GenomicsConverter code to utils-java and I will eventually change it there to support both API and GRPC and depend on the version in utils-java but I did not want to split a complicated refactoring across 2 projects for now so I expect to submit GenomicsConverter changes in this PR.

pom.xml references my local version of utils-java v0.28: I will replace that before submitting.

### CODE REVIEW

**Comment by deflaux** [comment]
@iliat I'll leave it to @dionloy to do the full review, but here are a couple side notes:

* Regarding `src/main/scripts` and documentation in general, please keep adding more to the [Google Genomics Cookbook](http://googlegenomics.readthedocs.org/en/latest/use_cases/run_picard_and_gatk/index.html).  Our goal is to have any docs on github cover material for people who want to *read and/or modify the code*.  The docs on http://googlegenomics.readthedocs.org/en/latest/use_cases/run_picard_and_gatk/index.html should cover cases where folks want to *run the code*.  Please see how we are doing this for the various Dataflow pipelines.
* Regarding travis-ci and "Could not resolve dependencies for project com.google.cloud.genomics:gatk-tools-java:jar:1.2-SNAPSHOT: Failure to find com.google.cloud.genomics:google-genomics-utils:jar:v1beta2-0.28-SNAPSHOT" another thing you could do instead is reference utils-java `v1beta2-0.28` in pom.xml.  It doesn't exist yet, but the idea is that it will and when it does, this build will become green.  Then you don't need an extra commit to get rid of the `-SNAPSHOT` and just keep that change to pom.xml locally but not checked in.

File: src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+import htsjdk.samtools.TagValueAndUnsignedArrayFlag;
+import htsjdk.samtools.ValidationStringency;
+import htsjdk.samtools.util.StringLineReader;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Genomics Api based implementation.
```
Comment (dionloy) [inline src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java]: both this and the gRPC versions are "APIs".  Refer to this "Api" as the "REST" version instead.  So:

Genomics API:
  - REST
  - gRPC

Pckage name should be ...gatk.common.rest and ...gatk.common.api.

File: src/main/java/com/google/cloud/genomics/gatk/common/GenomicsConverterBase.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    CIGAR_OPERATIONS.put("ALIGNMENT_MATCH","M");
+    CIGAR_OPERATIONS.put("CLIP_HARD", "H");
+    CIGAR_OPERATIONS.put("CLIP_SOFT","S");
+    CIGAR_OPERATIONS.put("DELETE", "D");
+    CIGAR_OPERATIONS.put("INSERT", "I");
+    CIGAR_OPERATIONS.put("PAD", "P");
+    CIGAR_OPERATIONS.put("SEQUENCE_MATCH", "=");
+    CIGAR_OPERATIONS.put("SEQUENCE_MISMATCH", "X");
+    CIGAR_OPERATIONS.put("SKIP", "N");
+  }
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/gatk/common/GenomicsConverterBase.java]: Why duplicate it?  All this is already in `utils-java` in the following file:

https://github.com/googlegenomics/utils-java/blob/0b7a441c0611fed99f737ce073570a570b53c922/src/main/java/com/google/cloud/genomics/utils/ReadUtils.java

File: src/main/java/com/google/cloud/genomics/gatk/common/UnmappedReadsBase.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ */
+public abstract class UnmappedReadsBase<Read> implements UnmappedReads<Read> {
+  protected static final Logger LOG = Logger.getLogger(UnmappedReads.class.getName());
+  
+  /**
+   * Maximum number of reads we are prepared to keep in memory.
+   * If this number is exceeded, we will switch to the mode of ignoring
+   * unmapped mate pairs.
+   */
+  protected static final long MAX_READS = 100000000;
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/gatk/common/UnmappedReadsBase.java]: Any specific reason for this limit?  For some chromosomes this is 1/3 of the chromosome.  What would be the pros/cons as it ties with the rest of the implemented features of the client?

File: src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        read.getReadNumber() == 1) ? 128 : 0;  // second_in_pair
+    flags += (read.getSecondaryAlignment() != null 
+        && Boolean.TRUE.equals(read.getSecondaryAlignment())) ? 256 : 0; // secondary_alignment
+    flags += (read.getFailedVendorQualityChecks() != null &&
+    	Boolean.TRUE.equals(read.getFailedVendorQualityChecks())) ? 512 : 0;// failed_quality_check
+    flags += (read.getDuplicateFragment() != null && 
+    	Boolean.TRUE.equals(read.getDuplicateFragment())) ? 1024 : 0; // duplicate_read
+    flags += (read.getSupplementaryAlignment() != null &&
+    	Boolean.TRUE.equals(read.getSupplementaryAlignment())) ? 2048 : 0; //supplementary_alignment
+    record.setFlags(flags);
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java]: A lot of this is already in here in some similar form - can it be consolidated?:

https://github.com/googlegenomics/utils-java/blob/0b7a441c0611fed99f737ce073570a570b53c922/src/main/java/com/google/cloud/genomics/utils/ReadUtils.java

File: src/main/java/com/google/cloud/genomics/gatk/common/GenomicsConverterBase.java (hunk 2 above)
Comment (iliat) [inline src/main/java/com/google/cloud/genomics/gatk/common/GenomicsConverterBase.java]: @pgrosu Please read what I wrote in the PR request - I am aware of this but will deal with this in a subsequent PR: I don't want to deal with complicate refactoring and changes across 2 repos :)

File: src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java (hunk 4 above)
Comment (iliat) [inline src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java]: @pgrosu Same comment as above: I know and will take care of this in a subsequent PR after the refactoring is done.

File: src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java (hunk 1 above)
Comment (iliat) [inline src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java]: @dionloy Makes sense, will do. But you mean gatk.common.rest and ...gatk.common.grpc (not api), right ?

@deflaux Will add documentation, but I assume no harm in having these scripts together with the code as well ?
And will do pom.xml change.

File: src/main/java/com/google/cloud/genomics/gatk/common/GenomicsConverterBase.java (hunk 2 above)
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/gatk/common/GenomicsConverterBase.java]: @iliat Sorry Ilia, got into a bad habit lately of jumping directly to code these days.  Thanks for point me in the right direction - makes sense :)

Thx,
~p

File: src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java (hunk 4 above)
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java]: Yup, sorry got into a bad code-reading habit lately :)

File: src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java (hunk 1 above)
Comment (dionloy) [inline src/main/java/com/google/cloud/genomics/gatk/common/api/GenomicsConverter.java]: Right, .rest and .grpc for pkg names and descriptions.

**Comment by deflaux** [comment]
@iliat regarding documentation, if the scripts *are* the documentation, then they should eventually go away.  If the scripts are what the documentation would tell users to download and run, then they are fine.  This isn't a hard and fast rule, just trying to make it such that people who want to run a tool to *do* something don't need to wade through the piles of stuff we have on github to find what they are looking for.  The cookbook should just tell them how to do the task at hand all on one page - a recipe :-)

**Comment by iliat** [comment]
@dionloy Changed package names to rest & grpc, fixed pom.xml and merged. After this is in I will go and change utils-java ReadUtils to be able to deal in GenomicsConverter with both flavors (rest, grpc) and then will remove GenomicsConverter base + descendants from here. 
@deflaux The example.sh script I feel is right here, because it shows how to properly run the example code that is in this repo (similar to how we have instructions on how to run grpc in utils-java doc).
But! I will also update documentation in start-here and Picard repo, after I officially bump the version of this in the central repository and so there will be a spelled out recipe for running with grpc there.

File: src/main/java/com/google/cloud/genomics/gatk/common/grpc/GenomicsConverter.java [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+// This file differs from GenomicsConverter in the utils-java package 
+// ( com.google.cloud.genomics.utils) and it depends on HTSJDK 1.118.
+// utils-java can not yet be moved to depend on HTSJDK due to pending
+// work (https://github.com/samtools/htsjdk/pull/55).
```
Comment (dionloy) [inline src/main/java/com/google/cloud/genomics/gatk/common/grpc/GenomicsConverter.java]: Is this comment still valid ?

**Comment by dionloy** [comment]
LGTM Thanks!

**Pull request merged**